front/components/assistant/markdown/CodeBlockWithExtendedSupport.tsx

```diff
@@ -0,0 +1,104 @@
+import { IconButton, SparklesIcon, WrenchIcon } from "@dust-tt/sparkle";
+import mermaid from "mermaid";
+import React, { useContext, useEffect, useRef, useState } from "react";
+
+import { CodeBlock } from "@app/components/assistant/markdown/CodeBlock";
+import { MarkdownContentContext } from "@app/components/assistant/markdown/MarkdownContentContext";
+import { classNames } from "@app/lib/utils";
+
+const MermaidGraph: React.FC<{ chart: string }> = ({ chart }) => {
+  const graphRef = useRef<HTMLDivElement | null>(null);
+
+  useEffect(() => {
+    if (graphRef.current) {
+      mermaid.initialize({ startOnLoad: false });
+      graphRef.current.innerHTML = chart;
+      void mermaid.init(undefined, graphRef.current);
+    }
+  }, [chart]);
+
+  return <div ref={graphRef} className="mermaid"></div>;
+};
+
+export function CodeBlockWithExtendedSupport({
+  children,
+  className,
+  inline,
+}: {
+  children?: React.ReactNode;
+  className?: string;
+  inline?: boolean;
+}) {
+  const validChildrenContent = String(children).trim();
+
+  const [showMermaid, setShowMermaid] = useState<boolean>(false);
+  const [isValidMermaid, setIsValidMermaid] = useState<boolean>(false);
+  const { isStreaming } = useContext(MarkdownContentContext);
+
+  useEffect(() => {
+    if (isStreaming || !validChildrenContent || isValidMermaid || showMermaid) {
+      return;
+    }
+
+    void mermaid
```

Comment:
Can we avoid thenable? (Was probably already there but if we you don't mind refactoring.